src/FwdState.cc

```diff
@@ -793,20 +823,28 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
         if (originWantsEncryptedTraffic && // the "encrypted traffic" part
                 !peer->options.originserver && // the "through a proxy" part
                 !peer->secure.encryptTransport) // the "exclude HTTPS proxies" part
-            return establishTunnelThruProxy();
+
+            return advanceDestination("establish tunnel thru proxy", answer.conn, [this,&answer] {
+                establishTunnelThruProxy(answer.conn);
+            });
     }
 
-    secureConnectionToPeerIfNeeded();
+    secureConnectionToPeerIfNeeded(answer.conn);
 }
 
 void
-FwdState::establishTunnelThruProxy()
+FwdState::establishTunnelThruProxy(const Comm::ConnectionPointer &conn)
 {
     AsyncCall::Pointer callback = asyncCall(17,4,
                                             "FwdState::tunnelEstablishmentDone",
                                             Http::Tunneler::CbDialer<FwdState>(&FwdState::tunnelEstablishmentDone, this));
     HttpRequest::Pointer requestPointer = request;
-    const auto tunneler = new Http::Tunneler(serverConnection(), requestPointer, callback, connectingTimeout(serverConnection()), al);
+    const auto tunneler = new Http::Tunneler(conn, requestPointer, callback, connectingTimeout(serverConnection()), al);
+
+    // TODO: Replace this hack with proper Comm::Connection-Pool association
```

Comment:
Can you clarify for me why we need a pool here?
It sounds like the tunneler needs to be a pool owner. That is easy enough to implement since a pool is just a container class which may be empty at any time.

Reply:
The fwdPconnPool is used only for FwdState.cc but not for tunnel.cc, however both are using the Http::Tunneler. 
If the fwdPconnPool is used then the Http::Tunneler should update its statistics in the case of errors.
We have a similar flag for the HappyConnOpener the HappyConnOpener::allowPconn_ to tell to HappyConnOpener whether to use fwdPconnPool or not.

Maybe it is not difficult to fix this inconsistency, but I believe it is required a different PR for this.

Reply:
A proper solution is far from trivial. IIRC, we have even attempted to implement it during this or another recent project, and the changes quickly grew out of control. This TODO should be addressed in a dedicated PR.